### DISCUSSION TITLE
reaching inputs

### DISCUSSION THREAD

**Discussion by godbout** [godbout/kindaVim.blahblah#43] [Ideas]
i think kV is doing pretty well *once it reached* an input. in text inputs that support the macOS Accessibility you can have real Vim motions, in text inputs that don't we can impersonate some Vim motions through key remapping. in non-text inputs you can move around using the natural Vim motions you're used too.

but sometimes you first need to reach those inputs. for some apps Alfred is enough. for some others you may need to fiddle around with the mouse.

some apps are dedicated to handle this, like [Shortcat](https://shortcatapp.com) and [Vimac](https://github.com/dexterleng/vimac). i like Shortcat coz it's fast, and rather than having to wait for some letters to show up that you need to read and decipher, you can just type naturally and get what you want. but it's outdated and doesn't work with Electron shit apps. Vimac is great too, but a bit slow to my taste. and having to decipher the letters brings more friction.

so i'm wondering. should this be included in kV? should this be another app? or should i go back to sleep.

**Comment by godbout** [comment]
also i like apps that are small, clean, and do one and one thing very, very well. i'm a bit reluctant to add this to kV if that means adding a lot of options.

**Reply by zmre** [reply]
I'm pretty new to kV (customer as of yesterday) so I may not know some ways of using it yet, but here's how I look at it:

1. I want to keep my fingers on the keyboard as much as possible. Anytime I am forced to the mouse, I'm a bit sad.
2. kV lets me be efficient in text boxes and to use my preferred keys (vim rather than built-in emacs) which lets me move around more quickly. But it only lets me move around within an input box or a pane -- not between them.

To me, the fact that I can't move around within an app is disappointing. I was using Mail.app, for example, and figuring out that kV is most useful when I'm typing a message, but frustrating if I want to move around messages because I often have to use the mouse to get the focus somewhere useful (and it isn't obvious where the focus even is -- like in the message display pane or the message list pane).

So my two cents: it's still a single thing you're doing well if you allow navigating a bit more.  That said, I didn't know about Shortcat and Vimac so maybe pairing is the way to go. I'm going to give Vimac a whirl. In one of your screencast videos I think you used Vimac and I thought it was a hidden feature of kV that I hadn't stumbled into yet.  Also, my instinct was to try switching into vi mode, then to use capital HJKL (and I tried variations with ctrl and other modifiers) to try to move around within a window.  If that's supposed to work (with the fn key?), then it failed for me.

**Reply by godbout** [reply]
> 1. I want to keep my fingers on the keyboard as much as possible. Anytime I am forced to the mouse, I'm a bit sad.

😅️ i can empathize. i'm sad too! (a lot.)

> 2. kV lets me be efficient in text boxes and to use my preferred keys (vim rather than built-in emacs) which lets me move around more quickly. But it only lets me move around within an input box or a pane -- not between them.

that's correct. kindaVim lets you use Vim motions within input texts. you can also use most Vim motions between UI Elements, like in your list of emails in Mail. go up and down the list, select mails, delete them, undo, etc. it's only possible to go through one input text/pane to the other if macOS allows it itself. if macOS allows you to do it for example through the arrow keys, you can in kindaVim 1) with `hjkl` if the app/text input is in the Key Mapping Family (by doing key remapping) or 2) with `fn` + `hjkl` if the app/text input is in Auto/PGR/Electron (using Accessibility to calculate caret position, etc.) 

> To me, the fact that I can't move around within an app is disappointing. I was using Mail.app, for example, and figuring out that kV is most useful when I'm typing a message, but frustrating if I want to move around messages because I often have to use the mouse to get the focus somewhere useful (and it isn't obvious where the focus even is -- like in the message display pane or the message list pane).

i see, yes. those are effectively two different (big) things, hence they would(will?) be two different apps. this is why there's apps like Shortcat and Vimac. they do actually solve a different problem, and in that sense they're quite complimentary of kindaVim.

i had the same feeling as you, which is why i actually wrote this brain dump. kV is great ONCE you reach an input. but sometimes pressing `tab` or `shift tab` doesn't help enough, and you would want to reach the input in a better way. but that's unfortunately out of scope for kV (i've started building something that would do that, kinda like Shortcat. that's a whole other thing).
 
> In one of your screencast videos I think you used Vimac and I thought it was a hidden feature of kV that I hadn't stumbled into yet. 

oh. remember which one? maybe one of the first ones? i remember paying attention not to use it to avoid that confusion.

> Also, my instinct was to try switching into vi mode, then to use capital HJKL (and I tried variations with ctrl and other modifiers) to try to move around within a window. If that's supposed to work (with the fn key?), then it failed for me.

not sure i get you here, sorry. what do you mean by moving around within a window? like from one input to the other?

i'm gonna add something in the comment below. with some screencasts. (there's gonna be more explanation coming in the Wizard and Families Pane from the Preferences in kV23, to (hopefully) explain better the difference between Accessibility Strategy and Keyboard Strategy.)

**Reply by godbout** [reply]
> Also, my instinct was to try switching into vi mode, then to use capital HJKL (and I tried variations with ctrl and other modifiers) to try to move around within a window. If that's supposed to work (with the fn key?), then it failed for me.

so this is what `HJKL` do in an input where kV can read the text:

https://user-images.githubusercontent.com/121373/171686495-c2a4c71c-2173-4332-9aea-73bb71c80fdc.mp4

* `H` goes to the first visible line (indent and count respected) as it's supposed to
* `J` join lines as it's supposed to
* `K` should run a program to lookup what's under the cursor, it doesn't (actually that makes me think that i could implement this)
* `L` goes to the last visible line (indent and count respected) as it's supposed to

i guess there could be some stuff like "if you're in a text FIELD rather than a text AREA, where `HJKL` are not useful because a text field is only one line, then in that case move to another input text?". but that's super funky. like super not expected/consistent. i feel there's already "hacks" like the `fn` key to enforce the Key Mapping (Keyboard Strategy) that are already a bit funky, but that one would be over the top i think.

also, `HJKL` will not work in input text where the text is not readable. something like `gg` can be mapped grossly to `cmd up`, but for moves that play with the Screen rather than the Text, you have to have access to a lot of other info in order to place the caret properly. that's not possible if the text is not accessible. that includes Mail in the compose input. (actually that includes all mail composer, because they're not standard text areas, they're basically HTML instead.)

**Reply by godbout** [reply]
>  If that's supposed to work (with the fn key?), 

regarding `fn`, here is the problem it solves: sometimes you have apps with some custom UI where the text is readable on a per line basis. you want to use the Accessibility Strategy because this is where there's the most number of Vim moves, and the most precise. but you also want to be able to go from one line to the other. when it's the case, you can enforce the Keyboard Strategy with `fn`. so `fn j`, rather than "calculating where the caret should go within the input text", will just be rapped to the `down` key, and the app will do it's custom UI thing, moving from one line to the other. happens in Things for example in checklists, or in a lot of Electron apps.

https://user-images.githubusercontent.com/121373/171689460-564d5ed9-f5fe-457a-99f6-78a0272e9f79.mp4

**Reply by godbout** [reply]
hope that explains the scope of kV better. have a look combining it with Vimac. give it a few days and see. maybe actually Vimac is more what you want. if that's the case that's fine, let me know and i give you a refund. in any case feel free to send your thoughts.

**Reply by zmre** [reply]
Thanks for the long reply: yes, you're absolutely right about capital HJKL and their uses so no worries there. I've seen them mapped in non-standard ways to switch between tabs and navigate history in, for example, browser plugins, which is why I was trying it.

I have Vimac going right now. Not as snappy as I would like, but it definitely helps a lot. If you're building something similar: I also like the scroll area functionality, but would like to be able to target which scroll area.

Regarding the video where I saw the focus hints jump up: https://youtu.be/f8XRcu00GIA?t=80

Regarding the fn button: I haven't found a place where it works or is needed yet, but there are a lot of times when I wish I could use fn-j to map to the down arrow so I don't have to take my hands off the home keys. I might not be in vim normal mode at those times though. Could you make this work generally?

**Reply by godbout** [reply]
> I've seen them mapped in non-standard ways to switch between tabs and navigate history in, for example, browser plugins, which is why I was trying it.

ok, i would have more to say but i'm not really sure i get correctly what you mean. so just in case, from your sentence above, maybe what you're trying to achieve is this?

1. remap keys within macOS itself
![Screen Shot 2022-06-03 at 02 25 59](https://user-images.githubusercontent.com/121373/171702876-780ab7a5-2054-4b6c-8cf1-8f65f2d896f5.png)

2. use the passthrough option in kindaVim
![Screen Shot 2022-06-03 at 02 41 45](https://user-images.githubusercontent.com/121373/171703044-4cd56b46-2db1-47f2-b1bc-5dc82e83da16.png)

like this you could be in kV Normal Mode, press `shift cmd H` (or `option H` or something with either `cmd` or `option` in it) and macOS would pick it up. i do it all the time to change tabs, move my windows around, etc.
passthroughs are done with `cmd` or `option`, not `control` or `shift` because those keys are keys that Vim uses for its moves.

> I have Vimac going right now. Not as snappy as I would like, but it definitely helps a lot. If you're building something similar: I also like the scroll area functionality, but would like to be able to target which scroll area.
> 
> Regarding the video where I saw the focus hints jump up: https://youtu.be/f8XRcu00GIA?t=80

ah, you're right 😅️ that was one of the early video yeah. mistake. also i actually don't use Vimac anymore because i'm having the same issue as you, way too slow. Shortcat is faster (but the issue with Shortcat is that it hasn't been updated in a while, and crashes quite often. but the dev is back to working on it.) 

> Regarding the fn button: I haven't found a place where it works or is needed yet, but there are a lot of times when I wish I could use fn-j to map to the down arrow so I don't have to take my hands off the home keys. I might not be in vim normal mode at those times though. Could you make this work generally?

`fn-j` in kV Normal Mode is always gonna be translated by the `down` key. and macOS will act like if you pressed that `down` key, even if you're in Normal Mode.

if you mean remapping this while in Insert Mode (like, the normal macOS), that's gonna be out of scope also. kV brings Vim moves when you're in Vim mode. what you need then is a key remapper. there's many tools for this, like [Hammerspoon](https://github.com/Hammerspoon/hammerspoon) and [Karabiner-Elements](https://karabiner-elements.pqrs.org). with those you can remap anything you want under macOS. those tools are specialized for this.

**Reply by godbout** [reply]
@zmre i think in the coming weeks/months i'll have some very early build of the app running. i'm gonna start working on it because mainly i need it badly :D it'll be a paid app like kV, probably same price model, but like kV, it'll start with a free beta (well, in that case alpha) that will always be available. will ping you when i release.

**Reply by clouedoc** [reply]
@godbout I want a ping too! I wanted to switch between inputs but couldn't

**Reply by godbout** [reply]
@clouedoc it's coming along: https://twitter.com/wooshyapp
but not sure when. kV is the priority. and i'm not sure about betas. i may start directly with a license when i have something that works ok, and let users decide if they want more (although, i already have a good roadmap for it!)

**Comment by godbout** [comment]
sending something now.

**Comment by godbout** [comment]
https://user-images.githubusercontent.com/121373/176876561-f8195909-1fc5-4122-8c70-f50bce1532ae.mp4

**Reply by godbout** [reply]
Wooshy will be targeting UI elements through a whole UI search. can also use kV moves to reach the first, last, next, etc. elements.

**Reply by clouedoc** [reply]
this is beatiful

**Reply by godbout** [reply]
only beatiful, not beautiful? ☹️🥺️

**Reply by clouedoc** [reply]
My Mac crashed before I could correct my sentence.
This is absolutely incredible. Never mankind has been so close to perfectness.
Like the human reigns on the earth, kindaVimCorp will reign on our macs

**Reply by godbout** [reply]
lol. E-Corp.

**Reply by clouedoc** [reply]
Elliot would definitely have used kV

**Reply by clouedoc** [reply]
Elliot would definitely have used kV

**Reply by jannis-baum** [reply]
This is everything I've ever wanted for macOS🥹 Is this already released or are you planning on releasing it for kV?

**Reply by godbout** [reply]
alpha started a few days ago!

https://wooshy.app
https://www.youtube.com/channel/UCFi6Z79gzGDdgw9fuEg7g4w
https://github.com/godbout/Wooshy.docs

**Reply by godbout** [reply]
@zmre see the links above, maybe this will fit better than kV for what you were mentioning in your first post.

**Comment by zmre** [comment]
Been using whooshy for a few days now. Not natural to reach for it yet, but slowly figuring out when it's useful and I like it. Adding ways to figure out the names of targets without words will be good -- but maybe not on hover but rather a nearby label?

The one thing that trips me up regularly: I look in one place and want to jump right to it, hit the hot key, but it turns out my foreground app is something else. So I exit out of the whooshy prompt and fumble around with cmd-tab or raycast to get to the proper window and then re-trigger whooshy. I know you have an alfred workflow for this (and raycast more or less does the trick), but it's a lot of extra steps when I really just want to jump directly somewhere.

As a suggestion, would it be possible to have a keystroke that allows me to expand the searched targets beyond the current app?  I'd imagine it something like this: I hit cmd-shift-space and start typing my target. I realize I'm in an idiot and my search was contrained to a different window than the one I was looking at. I hit a hotkey or add a special character to my search (let's say "#" for argument's sake, but I don't really care): then the search expands across all visible windows on all screens and highlights the hits so I can get where I want.  That would be tremendous.

The other suggestion I have so far -- and this one is much lower priority -- is a way to avoid tabbing through results. Most of the time I can search so I have zero or minimal tabbing, but a few times I've had to repeatedly bang on tab or ctrl-n to get where I want and when that happens I wish for the label overlays of other tools. I'd envision maybe typing my search and then hitting, let's say, ctrl-enter or cmd-enter and then instead of choosing the first option, it labels the selections and lets me one or two letter type to direct jump where I want to go.

I'm already impressed by the speed and so far it seems stable.  A big improvement over Vimac.  I only just noticed that the Vimac author is putting forward a new tool called Homerow, which I haven't tried.  But looking at that reminded me of a feature of Vimac that I did sometimes use, which was the ability to scroll various scrollable windows using the keyboard. I use vi browser plugins so that I can already do this in browsers where I most often scroll already, so not a huge need, but it did come in handy at times. So maybe a third suggestion there. I know you can drop into kindaVim and use j/k as up/down substitutes. Maybe that's good enough? Or maybe it would be nice to be able to select the scroll area when there are a few. For example, I usually have Slack open and have a scroll area for channels as well as for messages and that's an app where I often have to break out the mouse.

Anyway, great work, Whooshy is already pretty great and I haven't hit any bugs as yet that I've noticed.

**Reply by godbout** [reply]
> Adding ways to figure out the names of targets without words will be good -- but maybe not on hover but rather a nearby label?

theory sounds good, practice i doubt so. your screen is gonna be a mess. pollution everywhere.

the idea with the hover is not to be used once for each item eternally. once you know for example that reaching your list of accounts on Twitter is `account`, then you got it. next time you type `acc` and voilà 😎️

but still discussable. but i'm gonna answer the rest first, because it's gonna be related.

> I realize I'm in an idiot

we've all been there ☹️

> As a suggestion, would it be possible to have a keystroke that allows me to expand the searched targets beyond the current app? I'd imagine it something like this: I hit cmd-shift-space and start typing my target. I realize I'm in an idiot and my search was contrained to a different window than the one I was looking at. I hit a hotkey or add a special character to my search (let's say "#" for argument's sake, but I don't really care): then the search expands across all visible windows on all screens and highlights the hits so I can get where I want. That would be tremendous.

it's in the ROADMAP: https://github.com/godbout/Wooshy.docs#maybe
although 10% of probability that this will happen, at the time of writing (of the ROADMAP). since then i played a bit with it. i'd rate the probability at 1% now.

again, theory sounds good, but when you start getting your hands dirty, well, you see it's really dirty. what would need to be done—AT LEAST—is:
1. grab all visible windows. the API is expensive, but it's possible
2. grab ALL THE UI ELEMENTS of ALL VISIBLE WINDOWS. hmm hmm. Wooshy is fast now, for several reasons, one being not grabbing ALL THE UI ELEMENTS of ALL VISIBLE WINDOWS. 
3. once you have ALL THE UI ELEMENTS of ALL VISIBLE WINDOWS, you need to intersect all the windows lol. like you don't want to have hints for stuff that are hidden behind other windows. or do you? maybe actually you want a list of everything, not just the front window? (i'm thinking as i'm typing here.) if that's the case, a list of everything, yeah wrong project. there's something that may interest you though, see a bit later. edit: and if this is really what you want, imagine having a list of everything from every visible window. then when you activate a result, you will need this window to go upfront. so probably in the list you'll need to see which target belongs to which window etc. that seems to be like a MASSIVE FRICTION. like you're gonna spend time trying to decipher results before entering return. is there a possibility that you're thinking of what you're thinking now because you use full screen split or something similar?

> I'd envision maybe typing my search and then hitting, let's say, ctrl-enter or cmd-enter and then

hmm. someone hasn't read the Manual 😅️ https://github.com/godbout/Wooshy.docs#click

> The other suggestion I have so far -- and this one is much lower priority -- is a way to avoid tabbing through results. Most of the time I can search so I have zero or minimal tabbing, but a few times I've had to repeatedly bang on tab or ctrl-n to get where I want and when that happens I wish for the label overlays of other tools. I'd envision maybe typing my search and then hitting, let's say, ctrl-enter or cmd-enter and then instead of choosing the first option, it labels the selections and lets me one or two letter type to direct jump where I want to go.

so i'm gonna go for it again, have you read the [Manual](https://github.com/godbout/Wooshy.docs#manual)? you can add a _narrower_ to your search term. most of the time you'll end up with **very few** targets left. also you can reach the last one, first one, middle of the lists, etc, etc with the macOS standard keyboard shortcuts that you would expect. **EVEN BETTER**, if you're still a kV user, you can enter Normal Mode and use `jk`, `G`, `gg`, `CTRL-U`, etc, etc. (add Wooshy to the `Key Mapping` Family.) it's really, really fun :D

for the labelling/list of results, i'm thinking that what may better suit you is [Shortcat](https://shortcat.app). Vimac/Homerow does the letters hints, Wooshy does the typing, Shortcat does both. the philosophy is a bit different. Wooshy is for people who like to type and type fast. a big point of Wooshy also is to take as little screen estate as possible. you just wanna see your targets and reach them.

> I'm already impressed by the speed and so far it seems stable. A big improvement over Vimac. I only just noticed that the Vimac author is putting forward a new tool called Homerow, which I haven't tried.

i bought a license of Homerow. it's a great app. although if you have used Vimac, it's mostly the same, but faster.

> But looking at that reminded me of a feature of Vimac that I did sometimes use, which was the ability to scroll various scrollable windows using the keyboard. I use vi browser plugins so that I can already do this in browsers where I most often scroll already, so not a huge need, but it did come in handy at times. So maybe a third suggestion there. I know you can drop into kindaVim and use j/k as up/down substitutes. Maybe that's good enough? Or maybe it would be nice to be able to select the scroll area when there are a few. For example, I usually have Slack open and have a scroll area for channels as well as for messages and that's an app where I often have to break out the mouse.

yes. i have this in my internal ROADMAP of stuff to explore:
![Screenshot 2022-08-04 at 21 36 49](https://user-images.githubusercontent.com/121373/182860781-8ff1daa4-8ec2-4115-8a2c-6fdbaf5e9f87.png)

i wasn't planning to do it soon but i've been on Ventura for maybe one or two weeks and hell it's gonna be needed yeah. many places not accessible by keyboard anymore. need a mouse or trackpad. really shit. iOS style. anyways. i thought about it for a while, and the current conclusions would be: 1) not part of Wooshy 2) may be a part of kV, like you said. may need better handling of scroll areas etc like it's done currently for Safari. like you can do `jk`, `G`, `gg`, `CTR-D`, etc. that may require Wooshy to ping the scroll areas, which would be fine. it doesn't now because it's useless 3) if not in kV, probably a free, open source small app that just does this. i like and dislike this idea at the same time. i like it because i like my areas of responsibilities separated. Wooshy is not about this. kV can, it makes sense, which is why i would like to have it there. but a separate open source free app sounds good too. what i DONT LIKE about it is having another global keyboard shortcut. i really don't want this. so yeah, that's the current state of affairs.

> Anyway, great work, Whooshy is already pretty great and I haven't hit any bugs as yet that I've noticed.

then you're not using multiple monitors LOL

**Reply by zmre** [reply]
Oh I've definitely read the manual. I see now that suggesting ctrl-enter was bad. I haven't had cause to right click on a target as yet.

To your other solution: typing more to find the target isn't always the solution for me. For example, I've used Wooshy to select tabs in my browser. Sometimes the tab titles only show me the first two or three characters -- like of the subject of a gmail email or something -- and I just don't know what to type next. Sometimes I only see the icons and have basically nothing to go on. That's when the overlay shortcut approach is better. I don't really want a ton of overlays by default, but there are cases when it would be well worth it.

Regarding expanding to all visible UI elements: I get that can be a lot. Of course would only want visible elements. Would want it filtered. Again, I wouldn't want that to be the default mode, but something I could jump to if it might be quicker than exiting out of Wooshy, switching apps, and then reentering Wooshy and retyping.

I haven't used Shortcat, primarily because the website was down for months -- since you first mentioned it anyway. I do like the idea of typing to narrow targets though. Looking for the best of both worlds.

I'm definitely using multiple monitors. I guess I sometimes don't see the input box when I'm on the external monitor, but I generally don't look for that box. The search still works fine and I move to fast to bother looking at what Wooshy thinks I typed anyway.

On the scrolling thing: the bigger part of the ask is a way to put focus on a specific scroll area. Not sure the best way to do this, but windows frequently have multiple scroll areas. That's true right now as I type in an input box on Github. I don't have a great way to switch scroll areas right now so my scrolling (with arrows or j/k or whatever) moves to the right place.

**Reply by godbout** [reply]
> Oh I've definitely read the manual. I see now that suggesting ctrl-enter was bad. I haven't had cause to right click on a target as yet.

phew. glad i didn't write RTFM 😂️

> To your other solution: typing more to find the target isn't always the solution for me. For example, I've used Wooshy to select tabs in my browser. Sometimes the tab titles only show me the first two or three characters -- like of the subject of a gmail email or something -- and I just don't know what to type next. Sometimes I only see the icons and have basically nothing to go on. That's when the overlay shortcut approach is better. I don't really want a ton of overlays by default, but there are cases when it would be well worth it.

hmm but what if you knew the metadata for those icons? i guess you're referring to some same icons that you would keep using most of the time right?

> Regarding expanding to all visible UI elements: I get that can be a lot. Of course would only want visible elements. Would want it filtered. Again, I wouldn't want that to be the default mode, but something I could jump to if it might be quicker than exiting out of Wooshy, switching apps, and then reentering Wooshy and retyping.

still, to only show the visible UI elements, you have to first get all of them, and intersect. huge amount of power. gonna slow down things by a lot. i get that this wouldn't be the common case, like you would start without this, and activate if needed. personally if something like your scenario happens, yes, i use Alfred and switch window with the Wooshy: Window to the Foreground! (sometimes i wish i hadn't called it this way... long to type) and then blast with Wooshy again. in my case i'm ok with this. it's pretty case. but i get that it's some extra keys. not saying no, just thinking out loud and digesting.

> I haven't used Shortcat, primarily because the website was down for months -- since you first mentioned it anyway. I do like the idea of typing to narrow targets though. Looking for the best of both worlds.

the new Shortcat comes with both typing, and the hints. i think it's slower tho. may be worth having a look.

> I'm definitely using multiple monitors. I guess I sometimes don't see the input box when I'm on the external monitor, but I generally don't look for that box. The search still works fine and I move to fast to bother looking at what Wooshy thinks I typed anyway.

yeah, that's kind of the spirit :D no looking, just typing and enter.

> On the scrolling thing: the bigger part of the ask is a way to put focus on a specific scroll area. Not sure the best way to do this, but windows frequently have multiple scroll areas. That's true right now as I type in an input box on Github. I don't have a great way to switch scroll areas right now so my scrolling (with arrows or j/k or whatever) moves to the right place.

my first thought would be to have something in Wooshy that only targets the scroll area. maybe typing "scroll" or something? then you can navigate on those different areas, and choose one. then you use kV to move around. that's hypothetical for now. but i like this idea.

**Reply by godbout** [reply]
> my first thought would be to have something in Wooshy that only targets the scroll area. maybe typing "scroll" or something? then you can navigate on those different areas, and choose one. then you use kV to move around. that's hypothetical for now. but i like this idea.

had a go at it. doesn't work, at least with the new Ventura UI/UX system. need something done especially for that. not sure yet where it'll land: Wooshy, kV, or a separate thing.
